rust_icu_ubrk/src/lib.rs

```diff
@@ -0,0 +1,735 @@
+// by bindgen, so we define it ourselves here.
+pub const UBRK_DONE: i32 = -1;
+
+/// Rust wrapper for the ICU `UBreakIterator` type.
+pub struct UBreakIterator {
+    // Pointer to the underlying ICU representation.
+    rep: ptr::NonNull<sys::UBreakIterator>,
+
+    // The underlying C representation holds pointers to `text` and exactly one of
+    // {`locale`, `rules`, `binary_rules`} throughout its lifetime. We are
+    // responsible for ensuring that the pointers remain valid during that time,
+    // and for dropping the referenced values once the underlying C representation
+    // is released.
+    //
+    // A break iterator may be cloned, in which case the underlying C representation
+    // of the cloned iterator will hold pointers to the values pointed to by the
+    // original iterator, while maintaining its own internal iteration state.
+    //
+    // For these reasons, these fields are wrapped in `Rc`, ensuring that the
+    // referenced values (`text`, `locale`, etc.) are not released prematurely
+    // if the original break iterator is dropped before its clone. As break
+    // iterators are inherently not thread-safe [1], `Rc` was chosen over `Arc`.
+    //
+    // [1] https://unicode-org.github.io/icu/userguide/boundaryanalysis/#thread-safety
+    text: Rc<ustring::UChar>,
+    locale: Option<Rc<ffi::CString>>,
+    rules: Option<Rc<ustring::UChar>>,
+    binary_rules: Option<Rc<Vec<u8>>>,
+}
+
+// Implements `ubrk_close`.
+simple_drop_impl!(UBreakIterator, ubrk_close);
+
+impl Iterator for UBreakIterator {
+    type Item = i32;
+
+    /// Advances the break iterator's position to the next boundary after its
+    /// current position.
+    ///
+    /// Note that `ubrk_next` will _never_ return the first boundary. For example,
+    /// given a newly-initialized break iterator whose internal position is `0`,
+    /// the first invocation of `next` will return the _next_ boundary, not `0`.
+    /// If the caller requires the first boundary, it should utilize [`first`].
+    ///
+    /// Also note that interleaving calls to [`first`], [`last`], [`previous`],
+    /// [`preceding`], or [`following`] may change the break iterator's internal
+    /// position, thereby affecting the next value returned by `next`.
+    ///
+    /// Implements `ubrk_next`.
+    ///
+    /// [`first`]: #method.new
+    /// [`following`]: #method.following
+    /// [`last`]: #method.last
+    /// [`preceding`]: #method.preceding
+    /// [`previous`]: #method.previous
+    fn next(&mut self) -> Option<Self::Item> {
+        let index =
+            unsafe { versioned_function!(ubrk_next)(self.rep.as_ptr()) };
+        if index == UBRK_DONE {
+            None
+        } else {
+            Some(index)
+        }
+    }
+}
+
+impl UBreakIterator {
+    /// Reports the number of locales for which text breaking information is
+    /// available.
+    ///
+    /// Implements `ubrk_countAvailable`.
+    pub fn count_available_locales() -> i32 {
```

Comment:
I would replace this and the following methods with an iterator.
This avoids passing an index that is larger than the number of available locales.

Reply:
My intuition was to wrap these up as an iterator, so I feel somewhat validated to learn that you would have done the same. I think that I may have been a bit too literal-minded when reading the part of your [comment](https://github.com/google/rust_icu/pull/161#issuecomment-686054854) on my last PR related to leaving some aspects of API refinement to end users. Refactored in 70f4642.